File: 0118-pascals-triangle/0118-pascals-triangle.java
class Solution {
    public List<List<Integer>> generate(int numRows) {
        List<List<Integer>> pascalTriangle = new ArrayList<>();

        List<Integer> prevLevel = new ArrayList<>();
        prevLevel.add(1);
        
        for(int rowIdx = 0; rowIdx < numRows; rowIdx++){
            List<Integer> currLevel = new ArrayList<>();

            for(int idx = 0; idx <= rowIdx; idx++){
                int prevVal = idx > 0 ? prevLevel.get(idx - 1) : 0;
                int currVal = idx < prevLevel.size() ? prevLevel.get(idx) : 0;

                currLevel.add(prevVal + currVal);
            }

            prevLevel = currLevel;
            pascalTriangle.add(currLevel);
        }

        return pascalTriangle;
    }
}

File: 0118-pascals-triangle/0118-pascals-triangle.py
class Solution:
    def generate(self, numRows: int) -> List[List[int]]:

        pascal_triangle = []
        prev_level = [1]

        for row_idx in range(numRows):
            curr_level = [0] * (row_idx + 1)

            for idx in range(row_idx + 1):
                prev_val = prev_level[idx - 1] if idx > 0 else 0
                curr_val = prev_level[idx] if idx < len(prev_level) else 0

                curr_level[idx] = prev_val + curr_val
            
            prev_level = curr_level
            pascal_triangle.append(curr_level)
        
        return pascal_triangle
        